### ISSUE TITLE
re-use of the installer scripts?

### ISSUE THREAD

**Issue by frostworx** [lutris/lutris#3117]
Hi,
I'm the author of the Steam wrapper tool [SteamTinkerLaunch (or short stl)](https://github.com/frostworx/steamtinkerlaunch)
which is started directly as Steam startoption `stl %command%` and offers a wide range of extra features for the launched game.

I just wondered if it would be a good idea to re-use your "installer" database to optionally auto apply required steps
when a game is launched from inside steam (f.e. with the flag `AUTOLUTRIS=1`).

Similar features are already implemented, so an integration would be pretty seamless (see below).
I haven't found the possibility in your api to directly search for an installer by SteamID
_( f.e.: https://lutris.net/api/installers?search="$SteamAppId")_
so I'd like to ask if it would be ok for you if I'd implement (cautious) downloads directly from the api into my program to search there? If not (which would be fine for me of course) I guess the "installer" lines are not under a special License, so I could just reuse
the required steps to create so called "tweak files" (my implementation for auto-applying required steps) for my stl, right?

edit: If my quick&dirty test is correct this would be currently 215 steam games which would ootb and about 3300 more games with a installer script and a steamid, but not meant for steam (which could be used as well on demand - f.e.  `AUTOLUTRIS=2` per game)

**Comment by strycore** [comment]
This is some pretty important discussion here and raises a few things.

* Lutris installers are currently tied to getting the game files + setting up the game itself. We should separate those into 2 separate steps so we can easily reapply part of the installer at any point on an existing directory.
* We do not have Steam in the "Game Providers" on the website. This is a rather new feature and we currently have support for MAME, Humble and GOG. The thing is, in the lutris-collectors repo, I have started writing the code to get the steam games so the only missing part now is writing the import tasks on the website.
* So far Lutris has been dealing with Steam games by itself, without interacting with the website and just using the local files to determine Steam games. This is different from Humble and GOG where the user has to log in to the service to access the game library. This is because there is a Steam client but no Humble or GOG client.
* When we do have Steam in Game Providers on the website, we'll have every Steam game ever released. This will allow us to create game pages for all Steam games.
* Like we do for consoles, we can have "auto-installers" which would do nothing but install the game in Steam (a simple `game: appid: $APPID` installer), no extras steps applied. For that, someone would have to create an actual installer.
* I checked your script that injects stl into the Steam config and I'm glad it's possible to do so. If this is the case, then we could consider shipping stl as one of the runtime components, making it always available.
* I want to move away from a list of step to be applied on a game and start having a more descriptive approach. Like: "this game uses DX11 and .NET 4.7. This game also has quirks: it's incompatible with esync". Given this info, lutris would know how to do the right thing: apply the correct configuration, or install the right winetricks package.  This kind of metadata doesn't even have to be part of an installer, it could be kept elsewhere.
* Having an open API that would make game information like that available could be valuable for a lot of people.

There's going to be some major work done on anything Steam and Proton related for Lutris 0.5.9, this gives us time to get this thing correctly but also, it should be available soon-ish, since it would be the release after the one currently in dev.

**Comment by frostworx** [comment]
Hi there,
thanks for the comprehensive answer!
I already was very busy and I already have implemented (offline)

- download (re-using already downloaded data)
- merging everything into one parsible list
- filtering only steam relevant games into an extra list (pretty complicated sometimes, because of the structure of some entries so it could be improvable, a optional function also searching for "non steam related" tweaks might be implemented later)
- cutting out relevant data for every game into its own SteamAppID raw file
- parsing that raw file and extracting several relevant parts which build a new lutris autotweak tweakfile

- Following functions are currently imported into the tweakfles:
- winetricks packages
- commandline arguments
- env variables
- dll overrides
- pulseaudio
- xlive
- mfplat (wip, but almost done)

The user will be able to choose which platforms are used for autotweaks and in which order.
The autotweaks are loaded at first then, so global or other "self-rolled" tweak files override the settings done by the autotweaks, leaving the full control to the user.

When lutris is enabled, but no data were downloaded yet, everything will be setup at the first game launch, so it takes a bit longer. In this case only the tweakfile for the launched game will be created.
There's also the option to import autotweaks per commandline:
```
stl autotweaks PLAFTORM (optional steamid)'
will autogenerate all tweak files for every parsible game of platform PLATFORM or just for the optional SteamAppID
```
Currently (almost) implemented platforms are protonfixes and lutris and at least basic parsing of the proton issues is planned.

A decentralized open database which everybody can use easily and freely would be the best solution and imho should be in the interest of everybody providing a database with user contributed information. If you're planning to restructure/cleanup your database this would be awesome!

@ your
"Like we do for consoles, we can have "auto-installers" which would do nothing but install the game in Steam (a simple game: appid: $APPID installer), no extras steps applied. For that, someone would have to create an actual installer."

When a game uses stl and is initially started via
`steam steam://rungameid/$SteamAppID`
it will do everything required that the game works ootb without any user interaction.
So it would just work if you start it from lutris as well.

The only thing the user needs to do is to actually add stl to the game launch command.
This could be theoretically automated (like in my script [here](https://github.com/frostworx/steamtinkerlaunch#general-usage))
but I don't want to mess around with other peoples steam internal configs with batch editing it obviously.
Maybe adding `stl %command%` per game on demand would be a valid solution.

**Comment by frostworx** [comment]
Hi again, just released v1.6.0 with initial lutris support :)
https://github.com/frostworx/steamtinkerlaunch/releases/tag/v1.6.0
Not sure if you want to keep this issue open for further brainstorming, so *leaving it open*

**Comment by frostworx** [comment]
Just wanted to congratulate to your good and well comprehensible article and wish you all the best!
